### PR TITLE
UI change propose: mode name is assistant name

### PR DESCRIPTION
**Pull request by DanRioDev**
This PR resolves the inconsistent display of assistant mode labels in the session output window by leveraging the `mode` field already persisted by the OpenCode CLI in message JSON files. Previously, the client attempted to mutate `assistant_mode` in memory (which didn't persist across renders) and relied on flawed backfilling that corrupted historical messages. The new approach reads the stable `mode` field directly from disk, ensuring accurate, persistent labeling without client-side hacks.

**Key Changes:**
- Add `mode` to `Message` type and deprecate `assistant_mode`.
- Update display logic in `session_formatter.lua` to prioritize `message.mode`.
- Remove ineffective client-side mutation in `core.lua`.
- Maintain backward compatibility with a fallback chain.

**Impact:** All assistant messages now display their original agent/mode name (e.g., "BUILD") from the CLI, including historical ones. New messages show the current mode until saved.

**Tests:** All unit and minimal tests pass (`./run_tests.sh`). Verified with real session data showing `"mode": "neoagent"` in JSON files.

## Problem

Assistant messages in the session output were labeled generically as "ASSISTANT":
- Bad historical reading of MODE usage.

Root cause: The CLI persists a `mode` field (e.g., `"mode": "BUILD"`) in `~/.local/share/opencode/storage/message/{sessionID}/msg_*.json`, but the client ignored it.

Example of persisted CLI data (from a real message JSON):
```json
{
  "id": "msg_9a600eb280018HZQXM3pOjKc59",
  "role": "assistant",
  "mode": "BUILD",  // ← This is the stable field we now use
  "assistant_mode": null,  // ← Deprecated, always null from CLI
  "time": { "created": 1759426702120, "completed": 1759426724309 },
  // ... other fields
}
```

Before this fix, `session_formatter.lua` only checked `message.assistant_mode` (line 299), falling back to `state.current_mode` for the last message (lines 303-310). This worked for unsaved messages but corrupted history.

## Solution

1. **Type Update** (`lua/opencode/types.lua`):
   - Added `mode` field to `Message` class for the CLI-persisted agent name.
   - Marked `assistant_mode` as deprecated to signal future removal.

   **Before:**
   ```lua
   ---@class Message
   ---@field assistant_mode string|nil Assistant mode active when message was created
   ---@field error table
   ```

   **After:**
   ```lua
   ---@class Message
   ---@field mode string|nil Agent/mode used to create this message (from CLI)
   ---@field assistant_mode string|nil Assistant mode active when message was created (deprecated)
   ---@field error table
   ```

2. **Display Logic Fix** (`lua/opencode/ui/session_formatter.lua`):
   - Updated `_format_message_header` to prefer `message.mode` over `message.assistant_mode` (line 299).
   - Kept the fallback for the last message (unsaved) to use `state.current_mode`, but now it's only triggered if both `mode` and `assistant_mode` are missing.
   - Updated comment for clarity (line 296).

   **Before (lines 296-314):**
   ```lua
   -- Use the assistant_mode stored on the message only (stable label)
   local display_name
   if role == 'assistant' then
     local mode = message.assistant_mode
     if mode and mode ~= '' then
       display_name = mode:upper()
     else
       -- For the most recent assistant message, show current_mode if assistant_mode is missing
       local is_last_message = msg_idx == #state.messages
       if is_last_message and state.current_mode and state.current_mode ~= '' then
         display_name = state.current_mode:upper()
       else
         display_name = 'ASSISTANT'
       end
     end
   else
     display_name = role:upper()
   end
   ```

   **After (lines 296-314):**
   ```lua
   -- Use the mode field from the message (stable label from CLI)
   local display_name
   if role == 'assistant' then
     local mode = message.mode or message.assistant_mode
     if mode and mode ~= '' then
       display_name = mode:upper()
     else
       -- For the most recent assistant message, show current_mode if mode is missing
       -- This handles new messages that haven't been stamped yet
       local is_last_message = msg_idx == #state.messages
       if is_last_message and state.current_mode and state.current_mode ~= '' then
         display_name = state.current_mode:upper()
       else
         display_name = 'ASSISTANT'
       end
     end
   else
     display_name = role:upper()
   end
   ```

   This creates a robust fallback chain:
   - `message.mode` (CLI-persisted, stable for all messages) → Success: Shows "BUILD".
   - `message.assistant_mode` (deprecated, for legacy) → Fallback if no `mode`.
   - `state.current_mode` (only for last unsaved message) → Temporary display.
   - "ASSISTANT" (default).

3. **Remove Client-Side Mutation** (`lua/opencode/core.lua`):
   - Deleted the `after_run` mutation (lines 145-154) that set `last.assistant_mode = state.current_mode`.
   - This was ineffective (messages reload from disk) and caused backfill issues (fixed in prior commit `db5ab7b`).

   **Before (lines 142-156):**
   ```lua
   if state.windows then
     -- Persist assistant_mode on the latest assistant message if available
     local msgs = state.messages
     if msgs and #msgs > 0 then
       local last = msgs[#msgs]
       if last and last.role == 'assistant' then
         if not last.assistant_mode or last.assistant_mode == '' then
           last.assistant_mode = state.current_mode
         end
       end
     end
     ui.render_output()
   end
   ```

   **After (lines 142-146):**
   ```lua
   if state.windows then
     ui.render_output()
   end
   ```

## Expected Result

After merging, assistant message headers in the session output will display the correct, persisted mode name:

**Before (generic/incorrect):**
```
🧑 ASSISTANT (model name) (2 minutes ago)
> Assistant response here...
```

**After (accurate, from CLI):**
```
🧑 BUILD (model name) (2 minutes ago)
> Assistant response here...
```

- **Historical messages:** Show original `mode` (e.g., "PLAN" from opencode's JSON).
- **New messages (unsaved):** Show `state.current_mode` temporarily, then update to CLI `mode` on save.
- **User messages:** Unchanged ("USER"). // can later be opencode's username attribute from configs
- **Edge cases:** Fallback to "ASSISTANT" if no mode available; legacy `assistant_mode` supported.
- **Performance:** No change; reads existing JSON fields via `session.get_messages()`.

To test:
1. Open a session with mixed historical/new messages.
2. Verify labels match CLI modes (e.g., via `cat ~/.local/share/opencode/storage/message/{session}/msg_*.json | jq '.mode'`).
3. Switch modes and send a new message – label updates correctly on render.

### CODE REVIEW

File: lua/opencode/ui/session_formatter.lua [hunk 1]
```diff
@@ -49,6 +49,7 @@ function M.format_session(session)
     end
 
     if session.revert and session.revert.messageID == msg.id then
+      ---@type {messages: number, tool_calls: number, files: table<string, {additions: number, deletions: number}>}
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua:52]: The return annotation should be sufficient. no need to repeat this

File: lua/opencode/ui/session_formatter.lua [hunk 2]
```diff
@@ -109,7 +116,7 @@ end
 ---@param messages Message[] All messages in the session
 ---@param revert_index number Index of the message where revert occurred
 ---@param revert_info SessionRevertInfo Revert information
----@return {messages: number, tool_calls: number, files: {additions: number, deletions: number}[]}
+---@return {messages: number, tool_calls: number, files: table<string, {additions: number, deletions: number}>}
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: I rather keep the original type, it's easier to read

File: lua/opencode/ui/session_formatter.lua [hunk 3]
```diff
@@ -185,10 +192,10 @@ function M._format_revert_message(stats)
       end
       if #file_diff > 0 then
         local line_str = string.format(icons.get('file') .. '%s: %s', file, table.concat(file_diff, ' '))
-        local line_idx = M.output:add_line(line_str)
-        local col = #('  ' .. file .. ': ')
+        local line_idx = M.output:add_line(line_str) ---@type number
+        local col = #('  ' .. file .. ': ') ---@type number
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: No need for theses annotations,

File: lua/opencode/ui/session_formatter.lua [hunk 4]
```diff
@@ -185,10 +192,10 @@ function M._format_revert_message(stats)
...
         for _, diff in ipairs(file_diff) do
-          local hl_group = diff:sub(1, 1) == '+' and 'OpencodeDiffAddText' or 'OpencodeDiffDeleteText'
+           local hl_group = diff:sub(1, 1) == '+' and 'OpencodeDiffAddText' or 'OpencodeDiffDeleteText' ---@type string
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: No need for theses annotations,

File: lua/opencode/ui/session_formatter.lua [hunk 5]
```diff
@@ -270,14 +292,35 @@ function M._format_message_header(message, msg_idx)
 
   M.output:add_empty_line()
   M.output:add_metadata({ msg_idx = msg_idx, part_idx = 1, role = role, type = 'header' })
+
+  -- Use the mode field from the message (stable label from CLI)
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: No need for this comment

File: lua/opencode/ui/session_formatter.lua [hunk 6]
```diff
@@ -235,17 +251,22 @@ function M._format_patch(part)
...
+        display_line = restore_line,
+        range = { from = restore_line, to = restore_line },
       })
       M.output:add_action({
         text = '[R]estore file',
         type = 'diff_restore_snapshot_file',
         args = { part.hash },
         key = 'R',
+        display_line = restore_line,
+        range = { from = restore_line, to = restore_line },
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua:269]: All theses changes on  add_action are unreleated to the actual PR.

and the add_action already use the last line per  default.

Image

File: lua/opencode/types.lua [hunk 7]
```diff
@@ -256,6 +256,8 @@
 ---@field providerID string Provider identifier
 ---@field role 'user'|'assistant'|'system' Role of the message sender
 ---@field system_role string|nil Role defined in system messages
+---@field mode string|nil Agent/mode used to create this message (from CLI)
+---@field assistant_mode string|nil Assistant mode active when message was created (deprecated)
```
Comment (sudo-tee) [inline lua/opencode/types.lua]: Why deprecate something that did not exist in the first place ?

File: lua/opencode/types.lua [hunk 8]
```diff
@@ -256,6 +256,8 @@
 ---@field providerID string Provider identifier
 ---@field role 'user'|'assistant'|'system' Role of the message sender
 ---@field system_role string|nil Role defined in system messages
+---@field mode string|nil Agent/mode used to create this message (from CLI)
```
Comment (sudo-tee) [inline lua/opencode/types.lua]: No need for the (from CLI)

File: lua/opencode/ui/session_formatter.lua [hunk 9]
```diff
@@ -473,9 +519,12 @@ function M._format_tool(part)
   end
 
   local start_line = M.output:get_line_count() + 1
-  local input = part.state and part.state.input or {}
-  local metadata = part.state.metadata or {}
-  local output = part.state and part.state.output or ''
+   ---@type TaskToolInput|BashToolInput|FileToolInput|TodoToolInput|GlobToolInput|GrepToolInput|WebFetchToolInput|ListToolInput
+   local input = (part.state and part.state.input) or {}
+   ---@type ToolMetadataBase|TaskToolMetadata|WebFetchToolMetadata|BashToolMetadata|FileToolMetadata|GlobToolMetadata|GrepToolMetadata|ListToolMetadata
+   local metadata = (part.state and part.state.metadata) or {}
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: No need to type this the lsp is already detecting it

Image

File: lua/opencode/ui/session_formatter.lua [hunk 10]
```diff
@@ -503,7 +552,7 @@ function M._format_tool(part)
 
   M.output:add_empty_line()
 
-  local end_line = M.output:get_line_count()
+     local end_line = M.output:get_line_count() ---@type number
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: No need to type this and there is a weird indentation

File: lua/opencode/ui/session_formatter.lua [hunk 11]
```diff
@@ -532,7 +581,7 @@ function M._format_task_tool(input, metadata, output)
     end
   end
 
-  local end_line = M.output:get_line_count()
+     local end_line = M.output:get_line_count() ---@type number
```
Comment (sudo-tee) [inline lua/opencode/ui/session_formatter.lua]: No need to type this and there is a weird indentation

**Comment by sudo-tee** [comment]
Thanks for the PR

I like the idea of replacing the header.

However the PR needs a little cleaning as it touch a lot of unrelated stuff.

**Comment by DanRioDev** [comment]
Done!

**Comment by sudo-tee** [comment]
Thanks for the fixes
